src/Tickets/Ticket_Actions.php

```diff
@@ -0,0 +1,464 @@
+			// Not a ticket anymore...
+			return;
+		}
+
+		$event = $ticket->get_event();
+
+		if ( ! $event instanceof WP_Post || 1 > $event->ID ) {
+			// Parent event, no longer exists.
+			return;
+		}
+
+		$its_happening = true;
+
+		$prefix = $is_start ? 'start' : 'end';
+		$method = "{$prefix}_date";
+
+		$timestamp = $ticket->$method();
+
+		if ( ! $timestamp ) {
+			// No timestamp...
+			return;
+		}
+
+		if ( time() + 30 < $timestamp ) {
+			$its_happening = false;
+			$method        = "schedule_date_{$prefix}_action";
+			// The actual timestamp is not immediate. Lets reschedule closer to the actual event.
+			$this->$method( $ticket_id, $timestamp );
+		}
+
+		try {
+			/**
+			 * Fires when a ticket's start/end time is almost reached or reached or just reached briefly in the past.
+			 *
+			 * In your callbacks you should use the value of $its_happening to reliably determine if this event is going to be fired
+			 * again in the future or not. If $its_happening is false, the event will be fired again in the future otherwise it won't.
+			 *
+			 * @since TBD
+			 *
+			 * @param int     $ticket_id     The ticket ID.
+			 * @param bool    $its_happening Whether the event is happening or not.
+			 * @param int     $timestamp     The ticket start/end timestamp.
+			 * @param WP_Post $event         The event post object.
+			 */
+			do_action( "tec_tickets_ticket_{$prefix}_date_trigger", $ticket_id, $its_happening, $timestamp, $event );
+		} catch ( Exception $e ) {
+			do_action(
+				'tribe_log',
+				'error',
+				__( 'Ticket date action failed!', 'event-tickets' ),
+				[
+					'method'        => __METHOD__,
+					'error'         => $e->getMessage(),
+					'code'          => $e->getCode(),
+					'prefix'        => $prefix,
+					'timestamp'     => $timestamp,
+					'its_happening' => $its_happening,
+					'now'           => time(),
+				]
+			);
+
+			if ( did_action( 'action_scheduler_before_process_queue' ) ) {
+				/**
+				 * We are in AS context and AS expects to catch exceptions to mark an action as failed using the exception's message as the reason.
+				 */
+				throw $e;
+			}
+		}
+	}
+
+	/**
+	 * Syncs the action scheduler date actions.
+	 *
+	 * @since TBD
+	 *
+	 * @param int $ticket_id         The ticket ID.
+	 * @param int $start_timestamp   The ticket start date.
+	 * @param int $end_timestamp     The ticket end date.
+	 *
+	 * @return void
+	 */
+	protected function sync_action_scheduler_date_actions( int $ticket_id, int $start_timestamp, int $end_timestamp ): void {
+		if ( $start_timestamp >= $end_timestamp ) {
+			// What is going on ? We cant work with that...
+			return;
+		}
+
+		// Unschedule first pre-existing actions.
+		as_unschedule_action( self::TICKET_START_SALES_HOOK, [ $ticket_id ], self::AS_TICKET_ACTIONS_GROUP );
+		as_unschedule_action( self::TICKET_END_SALES_HOOK, [ $ticket_id ], self::AS_TICKET_ACTIONS_GROUP );
+
+		if ( time() > $end_timestamp ) {
+			// The ticket has already ended. Do nothing.
```

Comment:
"The ticket [sale] has already ended." - correct?
The **ticket** cannot really end.